hw/mcu/ambiq/apollo2/src/hal_timer.c

```diff
@@ -0,0 +1,786 @@
+    }
+
+    if (!nvic_configured) {
+        nvic_configured = 1;
+
+        NVIC_SetVector(CTIMER_IRQn, (uint32_t)apollo2_timer_isr);
+        NVIC_SetPriority(CTIMER_IRQn, (1 << __NVIC_PRIO_BITS) - 1);
+        NVIC_ClearPendingIRQ(CTIMER_IRQn);
+        NVIC_EnableIRQ(CTIMER_IRQn);
+    }
+
+    bsp_cfg = vcfg;
+    bsp_timer->cfg = *bsp_cfg;
+
+    return 0;
+}
+
+/**
+ * hal timer config
+ *
+ * Configure a timer to run at the desired frequency. This starts the timer.
+ *
+ * @param timer_num
+ * @param freq_hz
+ *
+ * @return int
+ */
+int
+hal_timer_config(int timer_num, uint32_t freq_hz)
+{
+#if !APOLLO2_TIMER_ANY_ENABLED
+    return SYS_EINVAL;
+#endif
+
+    struct apollo2_timer *bsp_timer;
+    uint32_t cont_cfg;
+    uint32_t once_cfg;
+    uint32_t sdk_cfg;
+    int rc;
+
+    bsp_timer = apollo2_timer_resolve(timer_num);
+    if (bsp_timer == NULL) {
+        return SYS_EINVAL;
+    }
+
+    rc = apollo2_timer_sdk_cfg(&bsp_timer->cfg, freq_hz, &bsp_timer->freq_hz,
+                               &sdk_cfg);
+    if (rc != 0) {
+        return rc;
+    }
+
+    /* Configure the continuous timer. */
+    cont_cfg = sdk_cfg | AM_HAL_CTIMER_FN_CONTINUOUS;
+    am_hal_ctimer_clear(bsp_timer->cont_timer_idx, AM_HAL_CTIMER_BOTH);
+    am_hal_ctimer_config_single(bsp_timer->cont_timer_idx, AM_HAL_CTIMER_BOTH,
+                                cont_cfg);
+
+    /* Configure the "once" timer. */
+    once_cfg = sdk_cfg | AM_HAL_CTIMER_FN_ONCE | AM_HAL_CTIMER_INT_ENABLE;
+    am_hal_ctimer_clear(bsp_timer->once_timer_idx, AM_HAL_CTIMER_BOTH);
+    am_hal_ctimer_config_single(bsp_timer->once_timer_idx, AM_HAL_CTIMER_BOTH,
+                                once_cfg);
+
+    /* Start the continuous timer. */
+    am_hal_ctimer_start(bsp_timer->cont_timer_idx, AM_HAL_CTIMER_BOTH);
+
+    return 0;
+}
+
+/**
+ * hal timer deinit
+ *
+ * De-initialize a HW timer.
+ *
+ * @param timer_num
+ *
+ * @return int
+ */
+int
+hal_timer_deinit(int timer_num)
+{
+#if !APOLLO2_TIMER_ANY_ENABLED
+    return SYS_EINVAL;
+#endif
+
+    struct apollo2_timer *bsp_timer;
+
+    bsp_timer = apollo2_timer_resolve(timer_num);
+    if (bsp_timer == NULL) {
+        return SYS_EINVAL;
+    }
+
+    if (bsp_timer->freq_hz == 0) {
+        /* Timer not enabled. */
+        return SYS_ENODEV;
+    }
+
+    am_hal_ctimer_stop(bsp_timer->cont_timer_idx, AM_HAL_CTIMER_BOTH);
+    am_hal_ctimer_stop(bsp_timer->once_timer_idx, AM_HAL_CTIMER_BOTH);
+    bsp_timer->freq_hz = 0;
+
+    return 0;
+}
+
+/**
+ * hal timer get resolution
+ *
+ * Get the resolution of the timer. This is the timer period, in nanoseconds
+ *
+ * @param timer_num
+ *
+ * @return uint32_t
+ */
+uint32_t
+hal_timer_get_resolution(int timer_num)
+{
+#if !APOLLO2_TIMER_ANY_ENABLED
+    return 0;
+#endif
+
+    const struct apollo2_timer *bsp_timer;
+
+    bsp_timer = apollo2_timer_resolve(timer_num);
+    if (bsp_timer == NULL) {
+        return 0;
+    }
+
+    return 1000000000 / bsp_timer->freq_hz;
```

Comment:
Could check whether hal_timer is enabled here, otherwise this'll generate div0 trap.

Reply:
Thanks - I have included this fix in the "Apollo2 - HAL timer implementation." commit.